### ISSUE TITLE
[Usage]: Eval pi05 RECAP policy on a real-world robot

### ISSUE THREAD

**Issue by sylvestf** [RLinf/RLinf#1135]
### Problem description

Hello,

Thanks for the great work! It's been very helpful.

I followed the [RECAP example](https://rlinf.readthedocs.io/zh-cn/latest/rst_source/examples/embodied/recap.html) to train a policy on a real-world task:
- SFT a pi0.5 policy with my real-world data
- Collected rollout data in LeRobot format
- Successfully ran RECAP training

Now I'm stuck at real-world evaluation. The RECAP model seems not to be compatible with the standard OpenPI inference server due to architecture changes. So what's the recommended way to evaluate a RECAP-trained policy on a real-world robot? Is there an existing evaluation script, or should I write a custom one?

Thanks for your help!

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by AiSaurabhPatil** [comment]
Hi @sylvestf  
I am exploring the pi0.5 RECAP method . 
I just wanted to know , how did you collected the rollout dataset ? 
did you setup the RL environment in any simulator for your usecase with custom reward modelling ? or directly recorded the SFT  policy inference on a real robot. 
I am stuck at this step, your guidance would be helpful. 

Thanks

**Comment by sylvestf** [comment]
> Hi [@sylvestf](https://github.com/sylvestf) I am exploring the pi0.5 RECAP method . I just wanted to know , how did you collected the rollout dataset ? did you setup the RL environment in any simulator for your usecase with custom reward modelling ? or directly recorded the SFT policy inference on a real robot. I am stuck at this step, your guidance would be helpful.
> 
> Thanks

Hi @AiSaurabhPatil ,
I'm working on a real-world robot setting. We wrote code for collecting rollout data using ROS2 and converting it to LeRobot format for our specific setup. I haven't tried it in any simulation but I see in the official RLinf documentation they seem to have already implemented this, just not sure if it has been released yet.

**Comment by QuanluZhang** [comment]
> ### Problem description
> Hello,
> 
> Thanks for the great work! It's been very helpful.
> 
> I followed the [RECAP example](https://rlinf.readthedocs.io/zh-cn/latest/rst_source/examples/embodied/recap.html) to train a policy on a real-world task:
> 
> * SFT a pi0.5 policy with my real-world data
> * Collected rollout data in LeRobot format
> * Successfully ran RECAP training
> 
> Now I'm stuck at real-world evaluation. The RECAP model seems not to be compatible with the standard OpenPI inference server due to architecture changes. So what's the recommended way to evaluate a RECAP-trained policy on a real-world robot? Is there an existing evaluation script, or should I write a custom one?
> 
> Thanks for your help!
> 
> ### Before submitting a new issue...
> * [x]  Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

hi @sylvestf , thanks for your interest in RLinf. We are working on making the evaluation of embodied models more unified and reusable. The evaluation part for RECAP will be released at the end of May.

**Comment by AiSaurabhPatil** [comment]
Hi @sylvestf , 
did you run(inferenced) the finetuned pi0.5 policy done using SFT data (human expert demo data) on your real robot to collect Rollout dataset ? 
How did you designed the reward function for rollout dataset? , Did you manually rewarded each episode during rollout data collection ?

**Comment by Nadeem752** [comment]
@sylvestf  I did the same . I ran a 22k SFT and then ran RECAP value model 10k and CFG 15k . what exact issue prevents you model from running ? in my case it runs with open pi server but performs worse than the 22k model .

**Comment by AiSaurabhPatil** [comment]
Hey @Nadeem752  , Did you recorded the Rollout datasets after SFT ?

**Comment by Nadeem752** [comment]
Yes. you can find my issue here -https://github.com/RLinf/RLinf/issues/1160

**Comment by AiSaurabhPatil** [comment]
Dataset 2: RL Rollouts (lora_rollouts)
Total episodes: 156
Positive episodes: 156 (episodes that appear in positive frames)
Negative episodes: 156 (episodes that appear in negative frames)
Total frames: 85,719
Positive frames: 29,481 (34.4%)
Negative frames: 56,238 (65.6%)
Advantage source: meta/advantages_fail800.parquet
Advantage continuous range: min=-0.5521, max=+0.3693, mean=0.0000
Sampling weight: 1.0

@Nadeem752 , I analyzed the dataset shared in RECAP blog using Lerobot Visualizer , I saw that in rollout dataset they a full negative output (target failed)episode or positive output (target achieve) episode which they used for next steps , instead of positive/negative frames in a episode that you mentioned above. 
what is the performance you were getting on SFT lora finetuned model for your particular task . let's say out of 10 attempt how many were success or failure ?

**Comment by Nadeem752** [comment]
46% sucess rate for fully finetuned model . LORA i dont remmeber as i had to delete LORA model becuase RECAP cannot finetune a LORA generate weights as they use JAX . The RECAP model has 0% sucess rate could be because of issue @sylvestf  mentioned . Did you examine there LIBERO dataset? fully negative means they are all marked as negative ? isnt that wrong ? as recap needs to calculate advantage per frame .

**Comment by AiSaurabhPatil** [comment]
<img width="1016" height="639" alt="Image" src="https://github.com/user-attachments/assets/436435c1-dc49-47e9-95bd-d1fa45b05d56" />

@Nadeem752 
yes I examine LIBERO 10 dataset , where i analyze the parquet files of the positive and negative episodes within rollout dataset , there I found that until the task is success or failure at the end all the other frames are getting -1 reward , if it is a success it gets +1 rewards which results to 0 and -300 for failure episodes. based on these observation I think they are separating positive and negative based on the episode outcome , not on the frame level. 
I am curious about ,  how you are deciding which frame is positive and vise verse in a particular episode ?

**Comment by Nadeem752** [comment]
I do the same , I increase it to -800 . That is step 1 ,In step 3 you compute advantage for frame level . This is what I understood is the method behind  RECAP . Maybe I understand it incorrectly :/

**Comment by AiSaurabhPatil** [comment]
@Nadeem752  Did you finetuned the lerobot implemented pytorch version of pi0.5 policy using SFT dataset ?

**Comment by Nadeem752** [comment]
no the open pi version

**Comment by AiSaurabhPatil** [comment]
But the openpi implemented policy in jax  , how did you converted it into pytorch ?

**Comment by Nadeem752** [comment]
no only the LORA part -openpi now provides PyTorch implementations of π₀ and π₀.₅ models alongside the original JAX versions! The PyTorch implementation has been validated on the LIBERO benchmark (both inference and finetuning). A few features are currently not supported (this may change in the future):
